tests/h/services/group_test.py

```diff
@@ -62,6 +64,59 @@ def test_it_returns_None_if_no_matching_group(self, svc):
         assert svc.fetch_by_groupid("group:rando@dando.com") is None
 
 
+class TestFilter(object):
+    def test_it_filters_by_name(self, svc, groups):
+        filters = {"name": "Hello"}
+
+        filtered_groups = svc.filter(**filters)
+
+        assert len(filtered_groups.all()) == 1
+        assert filtered_groups[0].name == "Hello"
```

Comment:
I think these asserts can be reduced to one clearer assert by using a [matcher class](https://seanh.cc/post/matchers/). There's actually already a `GroupScopeWithOrigin` matcher class in this test file, you need a `GroupWithName` matcher that follows the same pattern. This also applies to the asserts in all the other tests in this class -- it looks like each test needs just one matcher-based assert. Once they're all using a single matcher-based assert it looks you might also be able to use `parametrize` to reduce them to a single test method.

```diff
diff --git a/tests/h/services/group_test.py b/tests/h/services/group_test.py
index 84582e1b5..9ddda4e90 100644
--- a/tests/h/services/group_test.py
+++ b/tests/h/services/group_test.py
@@ -70,8 +70,7 @@ class TestFilter(object):
 
         filtered_groups = svc.filter(**filters)
 
-        assert len(filtered_groups.all()) == 1
-        assert filtered_groups[0].name == "Hello"
+        assert filtered_groups.all() == [GroupWithName("Hello")]
 
     def test_it_returns_all_groups_if_no_filters(self, svc, groups):
         filtered_groups = svc.filter()
@@ -230,3 +229,15 @@ class GroupScopeWithOrigin(Matcher):
         if not isinstance(group_scope, GroupScope):
             return False
         return group_scope.origin == self.origin
+
+
+class GroupWithName(Matcher):
+    """Matches any Group with the given name."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __eq__(self, group):
+        if not isinstance(group, Group):
+            return False
+        return group.name == self._name
```

Reply:
I defer to you because of your expertise on testing, but I would say that I have encountered Matcher classes in other tests and I feel that they make tests slightly less readable and harder to decipher, as one has to go figure out the Matcher class to understand the intent of the test. It may well be inelegant to write

`assert something == "Hello"`

but it's unambiguous and easy to parse. Can you help me understand what you mean when you say it's a "clearer" assert with the Matcher class?

Your suggestion of asserting against `.all()` using a List is definitely smarter.

Reply:
Hopefully once you're familiar with the matcher class pattern, a matcher class's name alone will be enough to communicate what it tests. That's the intention anyway.

I think the problem is that in most of these tests you really want to test the contents of the `filtered_groups.all()` list - test that it contains these 3 items, that it contains all groups, test what order it's in, etc. And you end up having to do these awkward double asserts like:

```python
assert filtered_groups[0].name == "hello"
assert len(filtered_groups.all()) == 1
```

Or:

```python
assert filtered_groups[0].name == "Fingers"
assert filtered_groups[1].name == "Finger"
```

Or in `test_it_returns_all_groups_if_no_filters()` you do `assert len(filtered_groups.all()) == len(groups) + 1` which actually only tests the length of the list, not that it contains all the groups.

If this were a function that returned simple objects, say strings, rather than groups you would definitely be using single asserts like:

```python
assert filtered_groups.all() == ["first_expected_group", "second_expected_group"]
```

And wouldn't be doing multiple asserts or length checks. I think a single assert like this is clearer because the test only really wants to test one thing -- that it returned the two expected groups -- and now you have one assert that tests just that. And then since the form of the assert would be the same in every test, you could probably collapse all the test methods into one parametrized test.

But you can't do that because it returns `models.Group` objects and to do such an assert you would have to get your hands on the actual `Group`s in order to do:

```python
assert filtered_groups.all() == [first_expected_group_object, second_expected_group_object]
```

If you can't actually do that, then I think matchers are the next best thing, and by naming matcher classes well I think the code is quite readable:

```python
assert filtered_groups.all() == [GroupWithName("first"), GroupWithName("second")]
```

So the matcher class allows you to assert the length of the list, what groups it contains, and their order, in a single readable assert.

(In other cases you can also do things like `assert GroupWithName("foo") not in some_list` etc.)